### PR TITLE
Governance relaying fixes

### PR DESCRIPTION
**Pull request by tgflynn**
The main effect of these changes is to avoid relaying objects and votes during orphan processing, which could cause rate check failures and banning.

### CODE REVIEW

File: src/governance.cpp [hunk 1]
```diff
@@ -265,7 +266,6 @@ void CGovernanceManager::CheckOrphanVotes(CGovernanceObject& govobj, CGovernance
             fRemove = true;
         }
         else if(govobj.ProcessVote(NULL, vote, exception)) {
```
Comment (UdjinM6) [inline src/governance.cpp]: Could probably also remove `fRemove` and rewrite/squash both `if` blocks here to smth like:
```
if(pairVote.second < nNow || govobj.ProcessVote(NULL, vote, exception)) {
    mapOrphanVotes.Erase(nHash, pairVote);
}
```

Comment (tgflynn) [inline src/governance.cpp]: I don't really want to be making cosmetic changes, or other unrelated changes, until the syncing and forking issues are resolved.

**Comment by UdjinM6** [comment]
Removal of duplicate Relay in CheckOrphanVotes makes sense (vote should be relayed by ProcessVote anyway), however it's not clear how relaying votes/objects which were processed and verified by ProcessVote/AddGovernanceObject could lead to banning tbh.

**Comment by tgflynn** [comment]
I think it's the same effect as for the change you made to prevent relaying during syncing.

Anytime multiple objects or votes are sent which are signed by the same MN it can trigger the rate check limit.

**Comment by UdjinM6** [comment]
But can/should this really happen when sync is already finished? I mean having orphan votes means that they probably arrived before their object and they should be relayed as soon as their object is received and verified. So by the end of sync there should be no orphans really. Orphans in non-synced mode should also be resolved quite quickly. So why would this trigger anything like that?

Speaking of rate limit check, the closest possible situation I see is when sync was just finished and then new object signed by some mn (objects signed by which we already got during sync) arrives (before we see next `mindiff` blocks). To solve this we shouldn't be updating `mapLastMasternodeTrigger` at all during sync process probably so that it was fresh new after sync. But that's kind of unrelated to orphans...

**Comment by tgflynn** [comment]
There's more than one situation.  There are 2 types of orphan votes and then there are orphan objects.

Orphan objects are objects for which we don't have the signing masternode.  Orphan votes can lack either the signing masternode or the object voted on.

I don't know exactly how objects are propagating in the network.  I have seen at least one case in the logs where a node didn't receive an object, apparently because the inv request was lost due to a full queue and then received the object later during orphan processing of votes (which causes missing objects to be requested from the sending node).  But when it was received it was rejected by the object rate limit check.

I think it's risky to have any bulk processing, certainly of objects, but possibly also of votes that can inject multiple objects/votes into the network rapidly because some of them may be signed by the same masternode.

If you're right that we need to relay during orphan processing to ensure normal propagation through the network then I'm not sure how we can make the rate limits work.

**Comment by UdjinM6** [comment]
>because the inv request was lost due to a full queue

That would be a dangerous situation by itself but I know of no normal conditions when this could happen. invs are sent in bulk of up to 50000 items and all items that won't fit are simply sent in next inv. Or which queue are you referring to?

>But when it was received it was rejected by the object rate limit check.

So object was meant to be received on some block height but while it was sitting in orphans and we requested some peer for mn entry which signed it, new block arrived so next time another object arrived signed by the same mn after `mindiff` blocks (on sending peer) our node thought that it's still `mindiff - 1` and banned sending peer. Smth like that? I thought that 10% threshold in `MasternodeRateCheck` was meant to solve this but even if it doesn't help (which is odd, even on testnet it's 2 blocks which _should_ be more than enough) how would proposed changes solve this?

>If you're right that we need to relay during orphan processing to ensure normal propagation through the network then I'm not sure how we can make the rate limits work.

I still don't understand why you think that relaying of re-verified orphan objs/votes is broken. I agree though that rate limits are - the situation you described ("10% doesn't help") and the situation I described above ("new objs/votes right after initial sync") should be addressed in some way but these changes solve none of them imo.

**Comment by tgflynn** [comment]
The inv being dropped is not related to this PR but what I observed was a "got inv" message with fAlreadyHave false (main.cpp:5531) which was not followed by an askfor message for the same object hash.

I think the only things that are likely to cause that are one of the early returns here: https://github.com/dashpay/dash/blob/v0.12.1.x/src/net.cpp#L2447.

For the rest I will have to think more about how to respond to your questions.

Is it not clear that the orphan processing can cause multiple objects and/or votes from the same MN to be resent within a short time interval and hence trigger the rate limit ?

**Comment by UdjinM6** [comment]
>I think the only things that are likely to cause that are one of the early returns here: https://github.com/dashpay/dash/blob/v0.12.1.x/src/net.cpp#L2447.

Hmm, that's interesting...

> Is it not clear that the orphan processing can cause multiple objects and/or votes from the same MN to be resent within a short time interval and hence trigger the rate limit ?

Resending multiple objs/votes signed by the same MN shouldn't trigger the rate limit _normally_ - what is orphan for your node should already be known by network from other nodes so all/most of such messages should be just ignored... which is why sending reprocessed orphans doesn't really make much sense, you're right in that part actually :)

Though the question that bothered me was how your node got to the point where it has such conditions... and that's probably an attack vector which can be performed by the mn itself! Here is how: start node, connect to as many peers as you can, sign few objs like if you are a mn and relay them (to slightly fill orphan map of your peers with your objects), sign and relay your mnb - done, peers you were connected to will _first_ relay mnb (i.e. their peers won't think that it's orphan objs/votes when your peers will relay) and only _then_ they will reprocess objs/votes, relay them and finally get banned by their peers. Find new peers, repeat.

This can be fixed by either applying rate limit to orphans... or simply by not relaying reprocessed orphan objs/votes since it's already doesn't really help to propagate messages as I mentioned above. Ok, I think changes make sense now, sorry it took me so long :D

**Comment by tgflynn** [comment]
Revert f9a7ea3, no that fixed a major oversight.  Without it local rpc votes won't get relayed at all.

I'm not sure now about the rest of it though.

Your earlier objections raised a doubt.  What if a node receives a vote from some MN before the govobj.  The vote would be an orphan and never relayed.  If all the nodes peers saw it in this order the vote would be lost (ie. never propagated further).  

Now the object and the vote shouldn't be processed out of order by the originating node, unless it is syncing, but your PR turned off relaying during syncing so we can exclude that case.  But I wonder if the inventory system always preserves the order of messages.

**Comment by UdjinM6** [comment]
I though we reprocess orphan votes when their object is received, or do we? I also quickly scanned through code and... that attack (and original issue) would make sense if we would apply any ban score for violating rate limit but we don't?... Or am I missing smth again? 🙈

**Comment by tgflynn** [comment]
We do penalize nodes for sending us votes too fast but I don't really see how your attack vector would work, at least if we turn off relaying during orphan processing as this PR does.  Otherwise fast votes would be invalid and not relayed.

**Comment by UdjinM6** [comment]
Hmm.. actually no, not for votes too
```
exception = CGovernanceException(ostr.str(), GOVERNANCE_EXCEPTION_TEMPORARY_ERROR);
```
i.e. penalty is 0

**Comment by tgflynn** [comment]
Yes, your right, I must have looked at the wrong case.

Anyway the real problem this is causing is loss of objects, which leads to forks, not banning.

**Comment by UdjinM6** [comment]
>the real problem this is causing is loss of objects

Ah, so currently rate limit simply push objs to orphans which we reprocess...  and fail because of the very same rate limit, right?

re f9a7ea3: I think it's quite confusing to have 2 functions with the same name (`ProcessVote`), only one of which would also relay processed vote. I would instead call `vote.Relay()` in rpc explicitly.

**Comment by tgflynn** [comment]
That's a lot of places you have to call that function.  I think it's better to have it centralized.

I think it's generally easier to reason about code when you have the logic more centralized rather than spread out in a lot of different places.

As for the overloaded function names, only one of them is public.  But I'll change the name to ProcessVoteAndRelay.

**Comment by UdjinM6** [comment]
There are only 4 places and all of them are in rpc module. Also, performing kind of atomic operations like `create`, `sign`, `process` i.e. `validate`, `relay` doesn't really have to be centralized imo. Changing name to `ProcessVoteAndRelay` would work too though :)

**Comment by tgflynn** [comment]
> Ah, so currently rate limit simply push objs to orphans which we reprocess... and fail because of the very same rate limit, right?

We have orphans because sometimes we don't get objects when we should, maybe because the inventory system is not a fully reliable delivery system, like we have been assuming.

I don't know what the root cause of the rate check failure I observed (which did fork a node) was but I submitted this PR because I think anytime we bulk relay objects or votes we risk triggering the rate limits.  I don't know that this PR would fix the case I saw but I think it should help reduce the number of scenarios that can produce this kind of problem.
